Fixing the fit_deriv method for Voigt1D class

lingxz opened the pull request:
Switched the first two elements in the returned list of the fit_deriv method for the Voigt1D class. 

Closes #4144

lingxz commented:
My first pull request, hopefully it's done correctly!

embray commented on CHANGES.rst:
```diff
@@ -480,6 +480,8 @@ Bug fixes
   - The keyword ``acc`` (for accuracy) is now correctly accepted by
     ``Simplex``. [#3966]
 
+  - Switched first two elements of the returned list of fit_deriv in Voigt1D class. [#4144]
```
The changelog entry isn't necessary--the Voigt1D model is new in v1.1 which hasn't been released yet, so any bug fixes to it prior to release don't need to be mentioned here.

embray replied:
My bad also for putting the wrong tag in #4144 which may have confused things.

embray commented on astropy/modeling/functional_models.py:
```diff
@@ -818,8 +818,7 @@ def fit_deriv(cls, x, x_0, amplitude_L, fwhm_L, fwhm_G):
 
-        dyda = [constant * V / amplitude_L,
-                - constant * dVdx * 2 * sqrt_ln2 / fwhm_G,
+        dyda = [- constant * dVdx * 2 * sqrt_ln2 / fwhm_G, constant * V / amplitude_L,
```
Could you please also put both of these on separate lines like they were before?  It's more readable that way (just in the correct order now! :)

embray commented:
Please add a regression test based on the one given in https://github.com/astropy/astropy/issues/4144#issuecomment-139396624

The test should check that the fitted parameters are close to the expected ones, using `np.allclose` (there are several examples of this in the existing tests).  You should be able to confirm that before the fix this test fails, and that it passes after the fix.

Thanks!

lingxz commented:
Thanks for the help, especially regarding the test! I've deleted the change log and added the test, and the put the two elements in the list in separate lines.

embray commented on astropy/modeling/tests/test_functional_models.py:
```diff
@@ -148,3 +149,13 @@ def test_Scale_inverse():
+    voi_init = models.Voigt1D(amplitude_L=-1.0, x_0=1.0, fwhm_L=5.0, fwhm_G=5.0)
+    fitter = fitting.LevMarLSQFitter()
+    voi_fit = fitter(voi_init, xarr, yarr)
+    assert_allclose(voi_fit.param_sets,voi.param_sets)
```
This is probably going to fail over some silly PEP-8 nitpicks.  Make sure each comma has one space after it.  Looks good otherwise :)

embray commented on astropy/modeling/tests/test_functional_models.py:
```diff
@@ -148,3 +149,13 @@ def test_Scale_inverse():
     assert_allclose(m.inverse(m(6.789)), 6.789)
+
+
+def test_Voigt1D():
```
Ah, this also needs to be prepended with:
```python
@pytest.mark.skipif("not HAS_SCIPY")
```
making sure to also add `from ...tests.helper import pytest` to the imports in the module.

embray commented on astropy/modeling/tests/test_functional_models.py:
```diff
@@ -7,6 +7,7 @@
 from numpy.testing import assert_allclose, assert_array_equal
 from .. import models
 from astropy.coordinates import Angle
+from astropy.modeling import fitting
```
I realize this had `from astropy.coordinates import Angle` just above, but that's an oversight.  This should use relative imports:
```python
from ...coordinates import Angle
from .. import fitting
```
could you change both of those while you're at it?

lingxz commented:
@embray fixed the imports and spaces and now the tests have all passed:D

embray commented on astropy/modeling/functional_models.py:
```diff
@@ -818,8 +818,8 @@ def fit_deriv(cls, x, x_0, amplitude_L, fwhm_L, fwhm_G):
 
-        dyda = [constant * V / amplitude_L,
-                - constant * dVdx * 2 * sqrt_ln2 / fwhm_G,
+        dyda = [- constant * dVdx * 2 * sqrt_ln2 / fwhm_G,
```
I guess one last nag would be that the exception is in this `-` sign right here, since it's a unary operator not a binary operator.  Kinda looks funny floating off by itself.  But I'll fix that later it's really no big deal and it seems to at least pass the tests.

embray commented:
Thanks!
